### PR TITLE
[luci/pass] Support RmsNorm Fuse

### PR DESCRIPTION
**Pull request by seockho-kim**
This commit supports RmsNorm fuse in luci pass.

ONE-DCO-1.0-Signed-off-by: Seockho Kim seockho.kim@samsung.com

issue: https://github.com/Samsung/ONE/issues/13964
draft: https://github.com/Samsung/ONE/pull/13967

### CODE REVIEW

**Comment by hseok-oh** [comment]
@nnfw-bot test nncc-debug nncc-release tizen-gbs

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+public:
+  luci::CircleNode *_ifm = nullptr;
+  luci::CircleMul *_mul_pow = nullptr;
+  luci::CircleMean *_mean = nullptr;
+  luci::CircleAdd *_add_epsilon = nullptr;
+  luci::CircleRsqrt *_rsqrt = nullptr;
+  luci::CircleMul *_mul_input = nullptr;
+  luci::CircleMul *_mul_gamma = nullptr; // not used
+  luci::CircleAdd *_add_beta = nullptr;  // not used
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:78]: let's remove these and add when necessary.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  const_one->dtype(loco::DataType::FLOAT32);
+  const_one->rank(1);
+  const_one->size<loco::DataType::FLOAT32>(1);
+  const_one->at<loco::DataType::FLOAT32>(0) = value;
+  return const_one;
+}
+
+bool RmsNormPattern::matched()
+{
+  // Check pattern
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:100]: IMHO, this comment can be removed as it doesn't give particular help reading the code.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 1 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:78]: Okay, I'll fix it.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+  CHECK_OR_FALSE(_mul_pow);
+  CHECK_OR_FALSE(_mul_pow->x() == _ifm);
+  CHECK_OR_FALSE(_mul_pow->y() == _ifm);
+
+  assert(_mul_gamma == nullptr);
+  assert(_add_beta == nullptr);
+  assert(_const_gamma == nullptr);
+  assert(_const_beta == nullptr);
+
+  // create gamma(1.0) and beta(0.0)
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:116]: Q) why are these 1.0 and 0.0 ?
Maybe add some comments about these 1.0 and 0.0 will be updated in the future.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 2 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:100]: Okay, I'll update it.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+public:
+  luci::CircleNode *_ifm = nullptr;
+  luci::CircleMul *_mul_pow = nullptr;
+  luci::CircleMean *_mean = nullptr;
+  luci::CircleAdd *_add_epsilon = nullptr;
+  luci::CircleRsqrt *_rsqrt = nullptr;
+  luci::CircleMul *_mul_input = nullptr;
+  luci::CircleConst *_const_epsilon = nullptr;
+  luci::CircleConst *_const_gamma = nullptr; // use default gamma(scale) 1.0
+  luci::CircleConst *_const_beta = nullptr;  // use default beta(offset) 0.0
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:79]:
```suggestion
  luci::CircleConst *_const_gamma = nullptr;
  luci::CircleConst *_const_beta = nullptr;
```

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  CHECK_OR_FALSE(_mul_pow->x() == _ifm);
+  CHECK_OR_FALSE(_mul_pow->y() == _ifm);
+
+  assert(_const_gamma == nullptr);
+  assert(_const_beta == nullptr);
+
+  /* Current FuseRmsNormPass assumes no gamma(scale) and beta(bias).
+   But, RmsNorm kernel expects gamma and beta.
+   So, it creates default gamma(1.0) and beta(0.0).
+  */
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:114]:
```suggestion
  /*
   NOTE: Current FuseRmsNormPass assumes no gamma(scale) and beta(bias).
   But, RmsNorm kernel expects gamma and beta.
   So, it creates default gamma(1.0) and beta(0.0).
  */
```

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 4 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:79]: Okay, I'll update it.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 5 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:114]: I'll update it :)

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  CHECK_OR_FALSE(_mul_pow->x() == _ifm);
+  CHECK_OR_FALSE(_mul_pow->y() == _ifm);
+
+  assert(_const_gamma == nullptr);
+  assert(_const_beta == nullptr);
+
+  /* NOTE: Current FuseRmsNormPass assumes no gamma(scale) and beta(bias).
+   But, RmsNorm kernel expects gamma and beta.
+   So, it creates default gamma(1.0) and beta(0.0).
+  */
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:114]: plz match comment begin/end style. current codes usually used the first one.
```suggestion
  /*
   NOTE: Current FuseRmsNormPass assumes no gamma(scale) and beta(bias).
   But, RmsNorm kernel expects gamma and beta.
   So, it creates default gamma(1.0) and beta(0.0).
  */
```
or
```suggestion
  /* NOTE: Current FuseRmsNormPass assumes no gamma(scale) and beta(bias).
   But, RmsNorm kernel expects gamma and beta.
   So, it creates default gamma(1.0) and beta(0.0). */
```

Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:114]: Okay, I'll update it.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+#define CHECK_OR_FALSE(condition) \
+  if (not(condition))             \
+    return false;
+
+luci::CircleConst *make_const_one(loco::Graph *graph, float value)
+{
+  auto const_one = graph->nodes()->create<luci::CircleConst>();
+  const_one->dtype(loco::DataType::FLOAT32);
+  const_one->rank(1);
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:90]: these looks missing
```diff
   const_one->rank(1); 
+  const_one->dim(0) = 1;
+  const_one->shape_status(luci::ShapeStatus::VALID);
```

File: compiler/luci/pass/src/FuseRmsNormPass.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+} // namespace
+
+namespace luci
+{
+
+bool FuseRmsNormPass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  // default gamma(scale) 1.0 and beta(offset) 0.0
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseRmsNormPass.cpp:204]: Q) is this necessary here?

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 7 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:90]: > these looks missing
> 
> ```diff
>    const_one->rank(1); 
> +  const_one->dim(0) = 1;
> +  const_one->shape_status(luci::ShapeStatus::VALID);
> ```

Okay, I'll add it.
BTW, I just copied it from FuseInstanceNormPass.
Is it a bug?

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 8 above)
Comment (seockho-kim) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:204]: > Q) is this comment necessary here?

I'll delete it.

File: compiler/luci/pass/src/FuseRmsNormPass.cpp (hunk 7 above)
Comment (seanshpark) [reply compiler/luci/pass/src/FuseRmsNormPass.cpp:90]: Maybe... can produce some problems but I wonder why there wasn't any.

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Pull request merged**